pkg/gardener/shoot/extender/provider/provider.go

```diff
@@ -78,20 +74,40 @@ func NewProviderExtenderPatchOperation(enableIMDSv2 bool, defMachineImgName, def
 			provider.Workers = append(provider.Workers, *rt.Spec.Shoot.Provider.AdditionalWorkers...)
 		}
 
-		controlPlaneConf, infraConfig := overrideConfigIfProvided(rt, existingInfraConfig, existingControlPlaneConfig)
+		provider.Workers = sortWorkersToShootOrder(provider.Workers, shootWorkers)
 
-		workerZones, err := getNetworkingZonesFromWorkers(provider.Workers)
+		workerZonesFromRuntime, err := getNetworkingZonesFromWorkers(provider.Workers)
 		if err != nil {
 			return err
 		}
 
-		// final validation
-		if err = checkWorkerZonesMatchProviderConfig(rt.Spec.Shoot.Provider.Type, workerZones, controlPlaneConf, infraConfig, true); err != nil {
+		workerZonesFromShoot, err := getNetworkingZonesFromWorkers(shootWorkers)
+		if err != nil {
 			return err
 		}
 
-		provider.ControlPlaneConfig = controlPlaneConf
-		provider.InfrastructureConfig = infraConfig
+		zonesAdded := newZonesAdded(workerZonesFromShoot, workerZonesFromRuntime)
```

Comment:
You don't need to make double check here since when`zonesEqual == false` then `len(zonesAdded)` is always `!= 0 `
Propose just to leave only `zonesEqual` call here and create helper function `mergeNewZones` that will be used to create final list of zones


The result code should be more easy to understand

```suggestion
if zonesEqual(workerZonesFromShoot, workerZonesFromRuntime) {
			provider.ControlPlaneConfig = existingControlPlaneConfig
			provider.InfrastructureConfig = existingInfraConfig
		} else {

			mergedWorkerZones := mergeNewZones(workerZonesFromShoot, workerZonesFromRuntime)
```			
			


And below: 

```golang
func mergeNewZones(a, b []string) []string {
	merged := make([]string, len(a))
	copy(merged, a)

	for _, zone := range b {
		if !slices.Contains(merged, zone) {
			merged = append(merged, zone)
		}
	}

	return merged
}			

```

Reply:
The check if `zonesEqual(workerZonesFromShoot, workerZonesFromRuntime) || len(zonesAdded) == 0` is for the case when we remove workers (e.g. there are two workers one is using zone a, the second is using a, b, c - when we remove the second one the infrastructureConfig should be left as it is). I will consider removing the check (the regeneration of the `infrastructureConfig` will not be needed but the result will be correct)